haystack/components/builders/hypothetical_doc_query.py

```diff
@@ -0,0 +1,127 @@
+from haystack import Pipeline, component, Document
+from haystack.components.embedders.sentence_transformers_document_embedder import SentenceTransformersDocumentEmbedder
+from haystack.components.generators.openai import OpenAIGenerator
+from haystack.components.builders import PromptBuilder
+
+from typing import List, Dict, Any
+from numpy import array, mean
+
+
+@component
+class InstructLM:
+    """
+    Subcomponent to generate hypothetical documents using an instruction-following language model (InstructLM).
+    """
+
+    def __init__(self, instruct_llm, nr_completions):
+        self.prompt_builder = PromptBuilder(
+            template="""
+        Given a question, generate a paragraph of text that answers the question.
+        Question: {{question}}
+        Paragraph:
+        """
+        )
+        self.generator = OpenAIGenerator(
+            model=instruct_llm, generation_kwargs={"n": nr_completions, "temperature": 0.75, "max_tokens": 400}
+        )
+
+        self.pipeline = Pipeline()
+        self.pipeline.add_component(name="prompt_builder", instance=self.prompt_builder)
+        self.pipeline.add_component(name="generator", instance=self.generator)
+        self.pipeline.connect("prompt_builder", "generator")
+
+    def to_dict(self) -> Dict[str, Any]:
+        # ToDo
+        pass
+
+    def from_dict(cls, data: Dict[str, Any]) -> "InstructLM":
+        # ToDo
+        pass
+
+    @component.output_types(replies=List[str])
+    def run(self, query: str):
+        result = self.pipeline.run(data={"prompt_builder": {"question": query}})
+        return result["generator"]
+
+
+@component
+class HyDEBuilder:
+    """
+    Subcomponent to build the Hypothetical Document Embedding (HyDE) from the generated hypothetical documents.
+    """
+
+    def __init__(self):
+        self.embedder = SentenceTransformersDocumentEmbedder(
+            model="sentence-transformers/all-MiniLM-L6-v2", progress_bar=False
+        )
+        self.embedder.warm_up()
+
+    def to_dict(self) -> Dict[str, Any]:
+        # ToDo
+        pass
+
+    @classmethod
+    def from_dict(cls, data: Dict[str, Any]) -> "HyDEBuilder":
+        # ToDo
+        pass
+
+    @component.output_types(embedding=List[float])
+    def run(self, answers: List[str]):
+        embeddings = self.embedder.run([Document(content=answer) for answer in answers])
+        stacked_embeddings = array([doc.embedding for doc in embeddings["documents"]])
+        avg_embeddings = mean(stacked_embeddings, axis=0)
+        hyde_vector = avg_embeddings.reshape((1, len(avg_embeddings)))
+        return {"hypothetical_documents": answers, "hypothetical_embedding": hyde_vector[0].tolist()}
+
+
+@component
+class HypotheticalDocumentEmbedder:
+    """
+    Hypothetical Document Embeddings (HyDE)
+
+    Given a query, HyDE first zero-shot prompts an instruction-following language model to generate a "fake"
+    hypothetical document but that captures relevant textual patterns.
```

Comment:
It's not mentioned here that we actually fake multiple documents and calculate their mean embedding.

Reply:
updated